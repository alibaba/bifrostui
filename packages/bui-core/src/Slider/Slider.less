@import '~@bifrostui/styles/mixins/index.less';

.bui-slider {
  --line-color: var(--bui-color-primary);
  --line-bg-color: var(--bui-color-border-default);
  --tooltip-font-size: var(--bui-text-size-3);
  --tooltip-color: var(--bui-color-white);
  --tooltip-bg-color: var(--line-color);
  position: relative;
  width: 100%;
  height: 2px;
  padding: 19px 0;
  box-sizing: border-box;
  font-family: var(--bui-font-family);

  &-disabled {
    pointer-events: none;
    opacity: 0.5;
  }

  &-line {
    position: absolute;
    top: 50%;
    left: 0;
    height: 2px;
    transform: translateY(-50%);
    background-color: var(--line-color);
  }

  &-rail {
    position: absolute;
    top: 50%;
    left: 0;
    width: 100%;
    height: 2px;
    transform: translateY(-50%);
    background-color: var(--line-bg-color);
  }

  &-button {
    display: flex;
    position: absolute;
    justify-content: center;
    align-items: center;
    width: 26px;
    height: 26px;
    background-color: #fff;
    box-shadow: #ddd 0 0 10px;
    border-radius: 50%;
    transform: translate(-50%, -50%);
    font-size: var(--bui-title-size-3);
    cursor: grab;

    // 增加点击热区大小
    &::after {
      position: absolute;
      content: '';
      top: -10px;
      bottom: -10px;
      left: -10px;
      right: -10px;
    }

    &-0 {
      left: 0%;
      top: 50%;
      z-index: var(--bui-z-index-affix);
    }

    &-1 {
      left: 0%;
      top: 50%;
      z-index: var(--bui-z-index-affix);
    }
  }

  &-tooltip {
    position: absolute;
    top: -14px;
    width: 46px;
    height: 24px;
    line-height: 24px;
    font-size: var(--tooltip-font-size);
    border-radius: 24px;
    color: var(--tooltip-color);
    text-align: center;
    background-color: var(--tooltip-bg-color);
    transform: translate(-50%, -50%);
    z-index: var(--bui-z-index-popover);

    &::after {
      position: absolute;
      bottom: -10px;
      left: 0;
      right: 0;
      margin: 0 auto;
      content: '';
      width: 0;
      height: 0;
      border: 6px solid transparent;
      border-top: 6px solid var(--tooltip-bg-color);
    }

    &-hidden {
      display: none;
    }
  }
}
