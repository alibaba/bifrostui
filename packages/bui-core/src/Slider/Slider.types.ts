import { OverrideProps } from '@bifrostui/types';
import React from 'react';

export type SliderValue = number | [number, number];

export type SliderProps<
  D extends React.ElementType = 'div',
  P = {},
> = OverrideProps<
  {
    props: P & {
      /**
       * 设置初始取值，类型为number时，使用单滑块，类型为[number, number]时，使用双滑块
       * 当组件非受控时使用
       */
      defaultValue?: SliderValue;
      /**
       * 设置当前值，类型为number时，使用单滑块，类型为[number, number]时，使用双滑块
       * 当组件受控时使用
       */
      value?: SliderValue;
      /**
       * 最小值，默认为0
       */
      min?: number;
      /**
       * 最大值，默认为100
       */
      max?: number;
      /**
       * 步长，取值必须大于 0，并且可被 (max - min) 整除
       * 默认值 1
       */
      step?: number;
      /**
       * 提示气泡是否总是可见，默认为false，滑动时才可见
       */
      tipVisible?: boolean;
      /**
       * 自定义提示气泡，默认：(value) => value
       */
      tooltipRender?: (value: number) => React.ReactNode;
      /**
       * 是否禁用，默认false
       */
      disabled?: boolean;
      /**
       * 开始图标
       */
      startIcon?: React.ReactNode;
      /**
       * 结束图标
       */
      endIcon?: React.ReactNode;
      /**
       * 禁止开始图标和结束图标交错，默认false，可以交错
       */
      disableSwap?: boolean;
      /**
       * 改变时的回调
       */
      onChange?: (
        event: React.SyntheticEvent,
        data: {
          value: SliderValue;
          buttonIndex: number;
        },
      ) => void;
    };
    defaultComponent: D;
  },
  D
>;
