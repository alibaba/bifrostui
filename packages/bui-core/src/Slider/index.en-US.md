---
group: Basic input
name: Slider
---

# Slider

Slide the input bar to select a value within a given range.

## Code demonstration

### Basic usage

Use`value`and`onChange`control the sliding input bar.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState(18);
  const handleChange = (e, data) => {
    setValue(data.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider value={value} onChange={handleChange} />
    </Stack>
  );
};
```

### Maximum and minimum values

Use`min`and`max`slide to limit the value range of the input bar.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState(29);
  const handleChange = (e, data) => {
    setValue(data.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider value={value} min={10} max={60} onChange={handleChange} />
    </Stack>
  );
};
```

### Custom Step Size

Adopt`step`customizable step size,`step`it should be an integer multiple of max min.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState(0);
  const handleChange = (e, data) => {
    setValue(data.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider value={value} step={10} onChange={handleChange} />
    </Stack>
  );
};
```

### Always display prompt bubbles

Enable`tipVisible`ï¼Œthe prompt bubble will always be displayed, otherwise it will only be displayed when sliding.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState(29);
  const handleChange = (e, data) => {
    setValue(data.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider value={value} tipVisible onChange={handleChange} />
    </Stack>
  );
};
```

### Double slider

When`value`or`defaultValue`the type is`[number, number]`the dual slider mode will be enabled.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState<[number, number]>([0, 20]);
  const handleChange = (e, data) => {
    setValue(data.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider value={value} onChange={handleChange} />
    </Stack>
  );
};
```

### Prohibit interlocking of front and rear buttons

When using double sliders, it is possible to use`disableSwap`prohibit interlocking of front and rear buttons.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

export default () => {
  const [value, setValue] = useState<[number, number]>([0, 20]);
  const handleChange = (e, data) => {
    console.log('123', data.value);
    setValue(data.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider value={value} disableSwap onChange={handleChange} />
    </Stack>
  );
};
```

### Start/End icon

Adopt`startIcon`customizable front button icon, through`endIcon`customizable back button icon.

```tsx
import React from 'react';
import { Slider, SliderValue, Stack } from '@bifrostui/react';
import { StarFilledIcon, StarOutlinedIcon } from '@bifrostui/icons';

export default () => {
  const defaultValue: SliderValue = [20, 60];
  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider
        startIcon={<StarFilledIcon htmlColor="#ff335c" />}
        endIcon={<StarOutlinedIcon htmlColor="#00d68f" />}
        defaultValue={defaultValue}
      />
    </Stack>
  );
};
```

### Disable

Use`disabled`disable components.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider defaultValue={50} disabled />
    </Stack>
  );
};
```

### Bubble prompt formatting

Use`tooltipRender`customizable prompt bubbles.

```tsx
import { Slider, Stack } from '@bifrostui/react';
import React, { useState } from 'react';

const tooltipRender = (value) => {
  const temp = Math.floor(value);
  const hour = temp >= 10 ? temp : `0${temp}`;
  const min = value % 1 === 0 ? '00' : '30';
  return `${hour}:${min}`;
};

export default () => {
  const [value, setValue] = useState(24);
  const handleChange = (e, data) => {
    setValue(data?.value);
  };

  return (
    <Stack style={{ padding: '0 30px' }}>
      <Slider
        min={0}
        max={24}
        step={0.5}
        value={value}
        tooltipRender={tooltipRender}
        onChange={handleChange}
      />
    </Stack>
  );
};
```

## API

| attribute     | explain                                                                | type                                                                                 | Default value |
| ------------- | ---------------------------------------------------------------------- | ------------------------------------------------------------------------------------ | ------------- |
| defaultValue  | Set initial value, to be used when the component is uncontrolled       | number \| [number, number]                                                           | -             |
| value         | Set the current value to be used when the component is controlled      | number \| [number, number]                                                           | -             |
| min           | minimum value                                                          | number                                                                               | 0             |
| max           | Maximum value                                                          | number                                                                               | 100           |
| step          | Step size, the value must be greater than 0 and divisible by (max min) | number                                                                               | 1             |
| tipVisible    | Prompt whether bubbles are always visible                              | boolean                                                                              | false         |
| tooltipRender | Customize prompt bubbles                                               | (value: number) => React.ReactNode                                                   | -             |
| disabled      | Do you want to disable it                                              | boolean                                                                              | false         |
| startIcon     | Start icon                                                             | React.ReactNode                                                                      | -             |
| endIcon       | End icon                                                               | React.ReactNode                                                                      | -             |
| disableSwap   | Prohibit interleaving of start and end icons                           | boolean                                                                              | false         |
| onChange      | Callback when changing                                                 | (event: React.SyntheticEvent,data: {value: SliderValue;buttonIndex: number}) => void | -             |

## Style variables

| attribute               | explain                                 | Default value              | global variable                    |
| ----------------------- | --------------------------------------- | -------------------------- | ---------------------------------- |
| --line-color            | Line foreground scenery                 | --bui-color-primary        | -                                  |
| --line-bg-color         | Line background color                   | --bui-color-border-default | -                                  |
| --width                 | Sliding input bar width                 | 100%                       | --bui-slider-width                 |
| --height                | Sliding input bar height                | 2px                        | --bui-slider-height                |
| --padding               | Slide the inner margin of the input bar | 19px 0                     | --bui-slider-padding               |
| --button-width          | Sliding button width                    | 100%                       | --bui-slider-button-width          |
| --button-height         | Slide button height                     | 2px                        | --bui-slider-button-height         |
| --button-font-size      | Font size of sliding button             | --line-color               | -                                  |
| --button-border-radius  | Sliding button border rounded corner    | 50%                        | --bui-slider-button-border-radius  |
| --tooltip-font-size     | Prompt for bubble font size             | --bui-text-size-3          | -                                  |
| --tooltip-color         | Prompt bubble font color                | --bui-color-white          | -                                  |
| --tooltip-bg-color      | Prompt bubble background color          | --line-color               | -                                  |
| --tooltip-width         | Prompt bubble width                     | 46px                       | --bui-slider-tooltip-width         |
| --tooltip-height        | Prompt bubble height                    | 24px                       | --bui-slider-tooltip-height        |
| --tooltip-line-height   | Prompt bubble height                    | 24px                       | --bui-slider-tooltip-line-height   |
| --tooltip-border-radius | Prompt bubble border rounded corners    | 24px                       | --bui-slider-tooltip-border-radius |
