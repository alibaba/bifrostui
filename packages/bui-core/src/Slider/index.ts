export { default, default as Slider } from './Slider';
export * from './Slider.types';
