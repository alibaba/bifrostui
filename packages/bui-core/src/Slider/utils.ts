import { SliderValue } from './Slider.types';

/**
 * 格式化数据
 * @returns undefined | number[]
 */
export const formatValue = (value: SliderValue) => {
  if (value === undefined) return value;

  return Array.isArray(value) ? value : [value, 0];
};

/**
 * 从小到大排序
 */
export const sortValue = (v: [number, number]): [number, number] => {
  if (!v) return v;
  return v.slice().sort((a, b) => a - b) as [number, number];
};

/**
 * 是否为相同值
 */
export const isSameValue = (newValue: SliderValue, oldValue: SliderValue) =>
  JSON.stringify(newValue) === JSON.stringify(oldValue);
