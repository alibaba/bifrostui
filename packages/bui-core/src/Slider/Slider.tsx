import { GripperBarVerticalIcon } from '@bifrostui/icons';
import {
  isMini,
  useForkRef,
  useTouchEmulator,
  useValue,
  getBoundingClientRect,
} from '@bifrostui/utils';
import clsx from 'clsx';
import React, {
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import './Slider.less';
import { SliderProps, SliderValue } from './Slider.types';
import { formatValue, isSameValue, sortValue } from './utils';

const classes = {
  root: 'bui-slider',
  button: 'bui-slider-button',
  tooltip: 'bui-slider-tooltip',
  disabled: 'bui-slider-disabled',
};

const SLIDER_BUTTON = {
  FRONT: 0,
  END: 1,
};

const Slider = React.forwardRef<HTMLDivElement, SliderProps>((props, ref) => {
  const {
    className,
    defaultValue,
    value: valueProp,
    min,
    max,
    step,
    tipVisible,
    tooltipRender,
    disabled,
    startIcon,
    endIcon,
    disableSwap,
    onChange,
    ...others
  } = props;

  const initialDefaultValue = useRef(defaultValue);
  // 处理value/defaultValue都不传的异常场景
  const latestVal = useRef<any>(0);
  const shouldProtection =
    valueProp === undefined && defaultValue === undefined;
  const value = shouldProtection ? latestVal.current : valueProp;

  const [rootRef, setRootRef] = useState(null);
  const handleRef = useForkRef(ref, setRootRef);
  useTouchEmulator(rootRef);

  // Slider BoundingClientRect
  const sliderRect = useRef<Record<string, any>>();
  // 滑动开始距离文档X轴原点距离
  const touchStartPageX = useRef<number>();
  // 按钮在滑动开始时距离左边的距离
  const touchStartLeft = useRef<number>();
  // 记录操作按钮
  const buttonIndex = useRef(0);
  // 记录双滑块touchMove之前的值
  const beforeMoveValue = useRef([0, 0]);

  // 默认图标
  const defaultIcon = <GripperBarVerticalIcon htmlColor="#2e333e" />;
  // 根据初始化时的数据判断是否为双滑块
  const initialRange = useRef(
    (value !== undefined && Array.isArray(value)) ||
      (defaultValue !== undefined && Array.isArray(defaultValue)),
  );
  // 是否为双滑块
  const range = useMemo(() => {
    // defaultValue传state，且通过setState改变defaltValue类型，此时使用初始化时的range（initialRange）判断
    if (
      Array.isArray(defaultValue) &&
      !Array.isArray(initialDefaultValue.current)
    ) {
      return initialRange.current;
    }

    return (
      (value !== undefined && Array.isArray(value)) ||
      (defaultValue !== undefined && Array.isArray(defaultValue))
    );
  }, [value, defaultValue]);

  // 气泡提示显示与隐藏
  const [frontTooltipVisible, setFrontTooltipVisible] = useState(false);
  const [endTooltipVisible, setEndTooltipVisible] = useState(false);

  /**
   * 格式化value、defaultValue
   * @type undefined | number[]
   */
  const formattedValue = formatValue(value);
  const formattedDefaultValue = formatValue(defaultValue);

  /**
   * 使用格式化后的value、defaultValue初始化sliderValue，应始终使用triggerChange改变值
   * @type undefined | number[]
   */
  const [sliderValue, triggerChange] = useValue({
    value: formattedValue,
    defaultValue: formattedDefaultValue,
    onChange: (e, data) => {
      onChange?.(e, {
        value: data?.value,
        buttonIndex: buttonIndex.current,
      });
    },
    config: {
      name: 'Slider',
      state: 'value',
    },
  });

  // 内部状态值
  const internalValue = useMemo<[number, number]>(() => {
    return sliderValue === undefined
      ? [min, max]
      : (sliderValue as [number, number]);
  }, [sliderValue]);
  // 记录上一次值
  const preValueRef = useRef(internalValue[buttonIndex.current]);
  // 记录气泡值
  const tooltipRef = useRef(internalValue);

  // 最大最小值范围
  const scope = useMemo(() => Number(max) - Number(min), [max, min]);

  // 提示气泡展示逻辑
  const tooltipVisible = useMemo(() => {
    const startVisible =
      buttonIndex.current === SLIDER_BUTTON.FRONT && frontTooltipVisible;
    const endVisible =
      buttonIndex.current === SLIDER_BUTTON.END && endTooltipVisible;

    return tipVisible || (!tipVisible && (startVisible || endVisible));
  }, [tipVisible, frontTooltipVisible, endTooltipVisible]);

  // 着色条样式
  const lineStyle = useMemo(() => {
    const sortedValue = sortValue(internalValue);
    let width = `${((sortedValue[1] - sortedValue[0] - min) / scope) * 100}%`;
    // 单滑块模式但动态修改value为数组，此时按照数组最小值计算着色条宽度
    if (!range && Array.isArray(value)) {
      width = `${((sortedValue[0] - min) / scope) * 100}%`;
    }
    return {
      left: range ? `${((sortedValue[0] - min) / scope) * 100}%` : '0%',
      width,
    };
  }, [internalValue, scope, value]);

  const getLineWidth = useCallback(async () => {
    if (!rootRef) return;
    sliderRect.current = await getBoundingClientRect(rootRef);
  }, [rootRef]);

  useLayoutEffect(() => {
    // 处理小程序初始化获取不到dom的问题
    setTimeout(() => {
      getLineWidth();
    }, 100);
  }, [rootRef]);

  const onTouchStart = (e) => {
    if (disabled) return;
    beforeMoveValue.current = internalValue;
    tooltipRef.current = internalValue;
    // 补偿获取宽度，隐藏元素getBoundingClientRect获取不到dom信息
    if (!sliderRect.current.width) {
      getLineWidth();
    }
    // currentTarget只存在于事件触发到事件处理结束之间，须在异步操作之前保存下来
    const currentTarget = isMini ? e.mpEvent.currentTarget : e.currentTarget;
    touchStartPageX.current = e.touches[0].pageX;
    touchStartLeft.current = currentTarget?.offsetLeft;

    // 显示气泡提示
    if (buttonIndex.current === SLIDER_BUTTON.END) {
      setEndTooltipVisible(true);
    } else {
      setFrontTooltipVisible(true);
    }
  };

  const onTouchMove = (e) => {
    if (disabled) return;
    // 触碰点当前x坐标
    const currentX = e.touches[0].pageX;
    // X轴滑动距离
    const moveX = currentX - touchStartPageX.current;
    // 计算下一个值
    const posX =
      ((touchStartLeft.current + moveX) / sliderRect.current.width) * scope;
    // 兼容一位小数
    let nextPosition = step < 1 ? Math.round(posX * 10) / 10 : Math.round(posX);

    if (nextPosition <= 0) nextPosition = 0;
    if (nextPosition >= scope) nextPosition = scope;

    // 滑动最新值
    const nextValue = min + nextPosition;
    const inStep = nextValue % step === 0;

    // 滑动的最新值与上一次值不等时 & 符合当前步长，则更新sliderValue
    if (!isSameValue(nextValue, preValueRef.current) && inStep) {
      let latestValue: SliderValue = nextValue;
      preValueRef.current = nextValue;
      // 单滑块
      tooltipRef.current = [latestValue, 0];

      if (range) {
        latestValue = [...(beforeMoveValue.current as [number, number])];
        if (disableSwap) {
          // 滑动前按钮
          if (
            buttonIndex.current === SLIDER_BUTTON.FRONT &&
            nextValue >= latestValue[SLIDER_BUTTON.END]
          ) {
            latestValue[SLIDER_BUTTON.END] = nextValue;
          } else {
            latestValue[buttonIndex.current] = nextValue;
          }
          // 滑动后按钮
          if (
            buttonIndex.current === SLIDER_BUTTON.END &&
            nextValue <= latestValue[SLIDER_BUTTON.FRONT]
          ) {
            latestValue[SLIDER_BUTTON.FRONT] = nextValue;
          } else {
            latestValue[buttonIndex.current] = nextValue;
          }
          tooltipRef.current = latestValue;
        } else {
          latestValue[buttonIndex.current] = nextValue;
          // 气泡值不应sort排序
          tooltipRef.current = latestValue;
          latestValue = sortValue(latestValue);
        }

        preValueRef.current = latestValue[buttonIndex.current];
      }

      latestVal.current =
        defaultValue !== undefined ? formatValue(latestValue) : latestValue;

      triggerChange(e, latestVal.current);
    }
  };

  const onTouchEnd = () => {
    if (disabled) return;
    // 置为排序后的值
    tooltipRef.current = internalValue;
    // 隐藏左右气泡提示
    setFrontTooltipVisible(false);
    setEndTooltipVisible(false);
  };

  const renderButton = (index: number) => {
    const valuenow = internalValue[index];

    return (
      <div
        key={index}
        role="slider"
        className={clsx(classes.button, `${classes.button}-${index}`)}
        style={{
          left: `${((valuenow - min) / scope) * 100}%`,
        }}
        tabIndex={disabled ? -1 : 0}
        aria-valuemin={props.min}
        aria-valuenow={valuenow}
        aria-valuemax={props.max}
        aria-orientation="horizontal"
        onTouchStart={(event) => {
          buttonIndex.current = index;
          onTouchStart(event);
        }}
        onTouchMove={onTouchMove}
        onTouchEnd={onTouchEnd}
        onTouchCancel={onTouchEnd}
      >
        {index === SLIDER_BUTTON.FRONT
          ? startIcon || defaultIcon
          : endIcon || defaultIcon}
      </div>
    );
  };

  return (
    <div
      className={clsx(classes.root, className, {
        [classes.disabled]: disabled,
      })}
      ref={handleRef}
      {...others}
    >
      {/* 底色条 */}
      <div className={`${classes.root}-rail`} />
      {/* 着色条 */}
      <div className={`${classes.root}-line`} style={lineStyle} />

      {/* 滑动按钮 */}
      {renderButton(SLIDER_BUTTON.FRONT)}
      {range && renderButton(SLIDER_BUTTON.END)}

      {/* 气泡提示 */}
      <div
        className={clsx(classes.tooltip, {
          [`${classes.tooltip}-hidden`]: !tooltipVisible,
        })}
        style={{
          left: `${
            ((tooltipRef.current[buttonIndex.current] - min) / scope) * 100
          }%`,
        }}
      >
        {tooltipRender(tooltipRef.current[buttonIndex.current])}
      </div>
    </div>
  );
});

Slider.displayName = 'BuiSlider';
Slider.defaultProps = {
  min: 0,
  max: 100,
  step: 1,
  disableSwap: false,
  tipVisible: false,
  tooltipRender: (value) => value,
};

export default Slider;
