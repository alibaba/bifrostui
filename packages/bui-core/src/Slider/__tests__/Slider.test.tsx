import React from 'react';
import { act, fireEvent, isConformant, render, screen } from 'testing';
import Slider from '..';

describe('Slider', () => {
  const originalModule = jest.requireActual('@bifrostui/utils');
  const restApi = jest.requireActual('react');
  const rootClass = 'bui-slider';

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
    jest.useRealTimers();
    jest.clearAllMocks();
  });

  isConformant({
    className: rootClass,
    displayName: 'BuiSlider',
    Component: Slider,
  });

  it('should render with default properties', () => {
    const fakeFn = jest.fn();
    const { container } = render(<Slider value={10} onChange={fakeFn} />);
    const slider = container.querySelector(`.${rootClass}`);
    expect(slider).toMatchSnapshot();
  });

  it('minimum value should controled by `min` property', () => {
    const { container } = render(<Slider min={10} defaultValue={10} />);
    const tooltip = container.querySelector(`.${rootClass}-tooltip`);
    expect(tooltip).toHaveTextContent('10');
  });

  it('maximum value should controled by `max` property', () => {
    const { container } = render(<Slider max={90} defaultValue={90} />);
    const tooltip = container.querySelector(`.${rootClass}-tooltip`);
    expect(tooltip).toHaveTextContent('90');
  });

  it('should render default value by `defaultValue` property', () => {
    const { container } = render(
      <Slider min={20} max={40} defaultValue={30} />,
    );
    const slide = container.querySelector(`.${rootClass}-button-0`);
    expect(slide).toHaveStyle({ left: '50%' });
  });

  it('should render current by `value` property', () => {
    const { container } = render(<Slider min={20} max={40} value={30} />);
    const slide = container.querySelector(`.${rootClass}-button-0`);
    expect(slide).toHaveStyle({ left: '50%' });
  });

  it('tooltip should hidden', () => {
    const { container } = render(<Slider value={0} tipVisible={false} />);
    const tooltip = container.querySelector(`.${rootClass}-tooltip`);
    expect(tooltip).toHaveClass(`${rootClass}-tooltip-hidden`);
  });

  it('tooltip should be formatted by `tooltipRender`', () => {
    const { container } = render(
      <Slider
        value={90}
        tooltipRender={(value) => {
          return <div className="formatter-tips">{value}</div>;
        }}
      />,
    );
    const tooltip = container.querySelector(`.formatter-tips`);
    expect(tooltip).toHaveTextContent('90');
  });

  it('should be disabled', () => {
    const { container } = render(<Slider value={10} disabled />);
    const slider = container.querySelector(`.${rootClass}`);
    expect(slider).toHaveClass(`${rootClass}-disabled`);
  });

  it('should render range', () => {
    const { container } = render(<Slider value={[0, 18]} />);
    const handlers = container.querySelectorAll(`.${rootClass}-button`);
    expect(handlers[0]).toHaveStyle({ left: '0%' });
    expect(handlers[1]).toHaveStyle({ left: '18%' });
  });

  it('should render startIcon', () => {
    const fakeChange = jest.fn();
    const { container } = render(
      <Slider
        value={10}
        startIcon={<div className="icon-start">start</div>}
        onChange={fakeChange}
      />,
    );
    const icon = container.querySelector(`.icon-start`);
    expect(icon).toBeVisible();
  });

  it('should render endIcon', () => {
    const fakeChange = jest.fn();
    const { container } = render(
      <Slider
        value={[10, 20]}
        endIcon={<div className="icon-end">end</div>}
        onChange={fakeChange}
      />,
    );
    const icon = container.querySelector(`.icon-end`);
    expect(icon).toBeVisible();
  });

  it('should trigger touchmove event', async () => {
    render(
      <Slider
        defaultValue={0}
        tooltipRender={(value) => <div>tooltip {value}</div>}
      />,
    );
    await act(async () => {
      await jest.runAllTimers();
    });
    act(() => {
      const button = screen.getByRole('slider');
      const mousedown = new MouseEvent('mousedown', {
        cancelable: true,
        bubbles: true,
      });
      const mousemove = new MouseEvent('mousemove', {
        cancelable: true,
        bubbles: true,
      });
      const touchMove = jest.fn();
      button.addEventListener('touchmove', touchMove, false);

      button.dispatchEvent(mousedown);
      button.dispatchEvent(mousemove);

      expect(touchMove).toBeCalled();
    });
  });

  it('should trigger touchend event', async () => {
    render(
      <Slider
        step={2}
        defaultValue={0}
        tooltipRender={(value) => <div>tooltip {value}</div>}
      />,
    );
    await act(async () => {
      await jest.runAllTimers();
    });
    act(() => {
      const button = screen.getByRole('slider');
      const mousedown = new MouseEvent('mousedown', {
        cancelable: true,
        bubbles: true,
      });
      const mousemove = new MouseEvent('mousemove', {
        cancelable: true,
        bubbles: true,
      });
      const mouseup = new MouseEvent('mouseup', {
        cancelable: true,
        bubbles: true,
      });
      const touchEnd = jest.fn();
      button.addEventListener('touchend', touchEnd, false);

      button.dispatchEvent(mousedown);
      button.dispatchEvent(mousemove);
      button.dispatchEvent(mouseup);

      expect(touchEnd).toBeCalled();
    });
  });

  it('should trigger touchmove event with range mode', async () => {
    const pageX = 30;
    const totalWidth = 300;
    jest.resetModules();
    // mock getBoundingClientRect
    jest.doMock('@bifrostui/utils', () => ({
      ...originalModule,
      getBoundingClientRect: () => {
        return Promise.resolve({
          width: totalWidth,
        });
      },
    }));
    jest.doMock('react', () => ({
      ...restApi,
    }));

    const { default: FakeSlider } = await import('../index');

    render(<FakeSlider defaultValue={[0, 5]} />);
    await act(async () => {
      await jest.runAllTimers();
    });

    const [button1] = screen.getAllByRole('slider');

    fireEvent.touchStart(button1, {
      touches: [
        {
          pageX: 0,
          pageY: 0,
        },
      ],
      cancelable: true,
      bubbles: true,
    });
    fireEvent.touchMove(button1, {
      touches: [
        {
          pageX,
          pageY: 0,
        },
      ],
      cancelable: true,
      bubbles: true,
    });
    // 因为双滑块可交错，第一个按钮移动30的距离后，第二个按钮会在左边，即变成第一个按钮
    // left计算逻辑详见源码
    expect(button1.style.left).toBe(`5%`);
  });

  it('should not swap end button', async () => {
    const pageX = 30;
    const totalWidth = 300;
    jest.resetModules();
    // mock getBoundingClientRect
    jest.doMock('@bifrostui/utils', () => ({
      ...originalModule,
      getBoundingClientRect: () => {
        return Promise.resolve({
          width: totalWidth,
        });
      },
    }));
    jest.doMock('react', () => ({
      ...restApi,
    }));

    const { default: FakeSlider } = await import('../index');

    render(<FakeSlider defaultValue={[0, 5]} disableSwap />);
    await act(async () => {
      await jest.runAllTimers();
    });

    const [button1, button2] = screen.getAllByRole('slider');

    fireEvent.touchStart(button1, {
      touches: [
        {
          pageX: 0,
          pageY: 0,
        },
      ],
      cancelable: true,
      bubbles: true,
    });
    fireEvent.touchMove(button1, {
      touches: [
        {
          pageX,
          pageY: 0,
        },
      ],
      cancelable: true,
      bubbles: true,
    });

    expect(button2.style.left).toBe(`${(pageX / totalWidth) * 100}%`);
  });

  it('should not swap start button', async () => {
    const pageX = 60;
    const totalWidth = 300;
    jest.resetModules();
    // mock getBoundingClientRect
    jest.doMock('@bifrostui/utils', () => ({
      ...originalModule,
      getBoundingClientRect: () => {
        return Promise.resolve({
          width: totalWidth,
        });
      },
    }));
    jest.doMock('react', () => ({
      ...restApi,
    }));

    const { default: FakeSlider } = await import('../index');

    render(<FakeSlider defaultValue={[30, 60]} disableSwap />);
    await act(async () => {
      await jest.runAllTimers();
    });

    const [button1, button2] = screen.getAllByRole('slider');

    fireEvent.touchStart(button2, {
      touches: [
        {
          pageX: 0,
          pageY: 0,
        },
      ],
      cancelable: true,
      bubbles: true,
    });
    fireEvent.touchMove(button2, {
      touches: [
        {
          pageX,
          pageY: 0,
        },
      ],
      cancelable: true,
      bubbles: true,
    });

    expect(button1.style.left).toBe(`${(pageX / totalWidth) * 100}%`);
  });
});
