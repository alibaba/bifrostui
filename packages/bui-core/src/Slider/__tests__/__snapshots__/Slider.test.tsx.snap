// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render with default properties 1`] = `
<div
  class="bui-slider"
>
  <div
    class="bui-slider-rail"
  />
  <div
    class="bui-slider-line"
    style="left: 0%; width: 10%;"
  />
  <div
    aria-orientation="horizontal"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="10"
    class="bui-slider-button bui-slider-button-0"
    role="slider"
    style="left: 10%;"
    tabindex="0"
  >
    <svg
      class="bui-svg-icon"
      color="#2e333e"
      focusable="false"
      viewBox="0 0 96 96"
    >
      <path
        d="M34.4 30c1.2 0 2 .8 2 2v32c0 1.2-.8 2-2 2s-2-.8-2-2V32c0-1.2.8-2 2-2zm28 0c1.2 0 2 .8 2 2v32c0 1.2-.8 2-2 2s-2-.8-2-2V32c0-1.2.8-2 2-2zm-14 0c1.2 0 2 .8 2 2v32c0 1.2-.8 2-2 2s-2-.8-2-2V32c0-1.2.8-2 2-2z"
      />
    </svg>
  </div>
  <div
    class="bui-slider-tooltip bui-slider-tooltip-hidden"
    style="left: 10%;"
  >
    10
  </div>
</div>
`;
